Fix in-video quiz question issues with display, sizing, and alignment

karandikar opened the pull request:
This set of commits fixes issues with the question alignment when the video isn't the first component, question sizing/alignment when a video is resized (window resize or un-fullscreening) with a question open, and questions displaying inconsistently because of a css issue.

karandikar committed:
Fix in-video quiz question misalignment when video isn't first component

If a video isn't the first component in the unit, the in-video-quiz
questions appear incorrectly positioned on the page. We reposition
the questions to properly cover the video while it's paused.

karandikar committed:
Fix issue with in-video quiz problem display css property

Previously, questions would sometimes fail to display in the in-video quiz
xblocks. This was likely due to the "display" css property not changing
from "none" to "block" for the problem component. This would also trigger
issues with the videoState variable as well.

karandikar committed:
Fix in-video-quiz questions not resizing when video is resized

Previously, if the video transitioned from fullscreen to regular size
with a question open, or if the window was resized with a question open,
the question would not resize properly (the position, height, and/or width
would be incorrect). This commit checks for changes to the video dimensions
every 100ms when a question is open, and resizes the question if a change
is detected.

karandikar commented:
@stvstnfrd @caesar2164 Let me know if any changes need to be made here. These changes solve the problem, but I'm aware they're inelegant/inefficient; I'll continue looking for an event I could listen for to detect a fullscreen change instead of checking every 200 milliseconds.

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +59,12 @@ function InVideoQuizXBlock(runtime, element) {
+        knownVideoPosition = $('.tc-wrapper', video).position().top;
+        knownVideoHeight = $('.tc-wrapper', video).css('height');
+        knownVideoWidth = $('.tc-wrapper', video).css('width');
+    }
```
```javascript
function getVideoDimensions() {
    var videoPosition = $('.tc-wrapper', video).position().top;
    var videoHeight = $('.tc-wrapper', video).css('height');
    var videoWidth = $('.tc-wrapper', video).css('width');
    return [videoPosition, videoHeight, videoWidth];
}
var knownVideoDimensions = getVideoDimensions();
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -98,6 +125,23 @@ function InVideoQuizXBlock(runtime, element) {
+              // both our known size measurements and the size of the problem
+              currentVideoPosition = $('.tc-wrapper', video).position().top;
+              currentVideoHeight = $('.tc-wrapper', video).css('height');
+              currentVideoWidth = $('.tc-wrapper', video).css('width');
```
define these with the function commented above!

karandikar commented:
@caesar2164 Fixed, let me know if anything else needs to change!

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -25,6 +31,7 @@ function InVideoQuizXBlock(runtime, element) {
 
             if (studentMode) {
                 setUpStudentView(component);
+                setUpVideoDimensions();
```
let's just make this:
```javascript
knownVideoDimensions = getVideoDimensions();
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,17 @@ function InVideoQuizXBlock(runtime, element) {
+
+    function setUpVideoDimensions() {
+        knownVideoDimensions = getVideoDimensions();
+    }
```
let's delete this function

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -98,6 +126,19 @@ function InVideoQuizXBlock(runtime, element) {
+              if ((currentVideoDimensions[0] !== knownVideoDimensions[0]) ||
+                  (currentVideoDimensions[1] !== knownVideoDimensions[1]) ||
+                  (currentVideoDimensions[2] !== knownVideoDimensions[2])) {
+                    resizeInVideoProblem(problemToDisplay, video);
```
let's pass `currentVideoDimensions` as the third value to this function

ie:
```javascript
resizeInVideoProblem(problemToDisplay, video, currentVideoDimensions);
```

caesar2164 replied:
rewrite:

```javascript
resizeInVideoProblem(problemToDisplay, currentVideoDimensions);
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -61,14 +79,22 @@ function InVideoQuizXBlock(runtime, element) {
     }
+    
+    function resizeInVideoProblem(currentProblem, currentVideo) {
+        videoDimensions = getVideoDimensions();
```
let's have this function take a third var for the `videoDimensions`

ie:
```javascript
function resizeInVideoProblem(currentProblem, currentVideo, videoDimensions) {
```

and then we can just delete this line.

caesar2164 replied:
Actually why do we even pass `currentVideo`?

caesar2164 replied:
better rewrite:

```javascript
function resizeInVideoProblem(currentProblem, videoDimensions) {
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -61,14 +79,22 @@ function InVideoQuizXBlock(runtime, element) {
+    
+    function resizeInVideoProblem(currentProblem, currentVideo) {
+        videoDimensions = getVideoDimensions();
+        currentProblem.css({top: videoDimensions[0], height: videoDimensions[1], width: videoDimensions[2]});
```
just a quick readability cleanup:
```javascript
currentProblem.css({
    top: videoDimensions[0], 
    height: videoDimensions[1], 
    width: videoDimensions[2]
});
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -87,7 +113,9 @@ function InVideoQuizXBlock(runtime, element) {
                 if (isProblemToDisplay) {
                   problemToDisplay = $('.xblock-student_view', this)
                   videoState.videoPlayer.pause();
+                  resizeInVideoProblem(problemToDisplay, video);
```
as a consequence of changing the `resizeInVideoProblem` function let's make this read: 

```javascript
resizeInVideoProblem(problemToDisplay, video, knownVideoDimensions);
```

caesar2164 replied:
rewrite:

```javascript
resizeInVideoProblem(problemToDisplay, knownVideoDimensions);
```

karandikar commented:
@caesar2164 Made the necessary edits! Let me know what you think

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,13 @@ function InVideoQuizXBlock(runtime, element) {
+        videoWidth = $('.tc-wrapper', video).css('width');
+        return [videoPosition, videoHeight, videoWidth];
+    }
+
```
let's add a `videoDimensionsDiffer()` function like so:

```javascript
function videoDimensionsDiffer(currentVideoDimensions, knownVideoDimensions) {
    if (currentVideoDimensions.length !== knownVideoDimensions.length) { 
        return true;
    }
    for (var i = 0, i < knownVideoDimensions.length, i++) {
        if (knownVideoDimensions[i] !== currentVideoDimensions[i]) {
            return true;
        }
    }
    return false;
}
```

karandikar replied:
@caesar2164 done!

caesar2164 replied:
awesome

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -98,6 +125,19 @@ function InVideoQuizXBlock(runtime, element) {
+              // both our known size measurements and the size of the problem
+              currentVideoDimensions = getVideoDimensions();
+
+              if ((currentVideoDimensions[0] !== knownVideoDimensions[0]) ||
```
this conditional can now be:

```javascript
if (videoDimensionsDiffer(currentVideoDimensions, knownVideoDimensions)) {
    ...
}
```

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -11,6 +11,12 @@ function InVideoQuizXBlock(runtime, element) {
+
     // Interval at which to check for problems to display
     // Checking every 0.5 seconds to make sure we check at least once per actual second of video
     var intervalTime = 500;
```
lets rename this variable, now that we have multiple intervals...

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -98,6 +137,17 @@ function InVideoQuizXBlock(runtime, element) {
+            resizeIntervalObject = setInterval(function () {
+
+              // check if the size has changed from the previous state; if so, update
+              // both our known size measurements and the size of the problem
```
this comment can probably go away now that the code is easier to read...

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
         }
```
Doh, I missed this before, lets move this to here, so it only gets done once...

```javascript
if (studentMode) {
    knownVideoDimensions = getVideoDimensions();
}
```

caesar2164 replied:
thanks!

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -11,20 +11,27 @@ function InVideoQuizXBlock(runtime, element) {
 
             if (studentMode) {
                 setUpStudentView(component);
+                knownVideoDimensions = getVideoDimensions();
```
move down to location of other comment.

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
+        return [videoPosition, videoHeight, videoWidth];
+    }
+
+    function videoDimensionsDiffer(newMeasurement, oldMeasurement) {
```
looks great!

caesar2164 commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -31,6 +37,7 @@ function InVideoQuizXBlock(runtime, element) {
         });
 
         if (studentMode) {
+          knownVideoDimensions = getVideoDimensions();
```
awesomesauce

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
     }
 
+    function getVideoDimensions() {
+        videoPosition = $('.tc-wrapper', video).position().top;
```
This variable needs declared first.

karandikar replied:
yep!

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
 
+    function getVideoDimensions() {
+        videoPosition = $('.tc-wrapper', video).position().top;
+        videoHeight = $('.tc-wrapper', video).css('height');
```
This variable needs declared first.

karandikar replied:
Yep!

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
+    function getVideoDimensions() {
+        videoPosition = $('.tc-wrapper', video).position().top;
+        videoHeight = $('.tc-wrapper', video).css('height');
+        videoWidth = $('.tc-wrapper', video).css('width');
```
This variable needs declared first.

karandikar replied:
Yep! My bad

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
+        videoPosition = $('.tc-wrapper', video).position().top;
+        videoHeight = $('.tc-wrapper', video).css('height');
+        videoWidth = $('.tc-wrapper', video).css('width');
+        return [videoPosition, videoHeight, videoWidth];
```
`videoPosition` is an unexpected value to be returned from `getVideoDimensions`;
is there a more apt name we can use?

stvstnfrd replied:
Also, looking at how this data is actually used, I suspect a dictionary/object may be a more apt data structure.

```javascript
return {
    'top': position,
    'height': height,
    'width': width
};
```

That means when we're modifying the CSS below, it's as easy as:

```javascript
currentProblem.css(videoDimensions);
```

Of course, the `for` loop will need updated, but that should be simple enough:

```javascript
for (var key in knownVideoDimensions) {
    if (knownVideoDimensions.hasOwnProperty(key)) {
        if (knownVideoDimensions[key] !== currentVideoDimensions[key]) {
            return true;
        }
    }
}
return false;
```

This makes the code more obvious, by avoiding directly indexing into the array `arr[0]`, `arr[1]`, etc., as that's opaque.

karandikar replied:
This is a great idea, implemented it

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
+    }
+
+    function videoDimensionsDiffer(newMeasurement, oldMeasurement) {
+        if (newMeasurement.length !== oldMeasurement.length) {
```
When would this case occur?
Seems like only on the first run, because `oldMeasurement` (via `knownVideoDimensions`) is not initialized when being declared; does that seem right?

If so, could we just initialize it to some value above and drop this special handling here?

```javascript
var knownVideoDimensions = [0, 0, 0];
```

karandikar replied:
This case would never occur, since knownVideoDimensions is set right at the outset. Regardless, it's now unnecessary with the changes you've suggested since!

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -87,17 +124,26 @@ function InVideoQuizXBlock(runtime, element) {
           clearInterval(intervalObject);
           if (problemToDisplay) {
+            resizeIntervalObject = setInterval(function () {
+              currentVideoDimensions = getVideoDimensions();
```
Should this be declared as a local variable?

karandikar replied:
Done!

stvstnfrd commented on invideoquiz/public/js/src/invideoquiz.js:
```diff
@@ -50,6 +57,25 @@ function InVideoQuizXBlock(runtime, element) {
     }
 
+    function getVideoDimensions() {
+        videoPosition = $('.tc-wrapper', video).position().top;
```
Nitpick: I tend to avoid prefixing variables if we're doing it for everything.
From the context of this function `getVideoDimensions`, it's obvious that
any dimensions we list are for videos, unless otherwise noted.

Not only does it cut down on some verbosity, but it also makes it easier to grep the code: grepping for `position` (case sensitive) won't find `videoPosition`, but it will find `position`.

karandikar replied:
Yeah, this is fair. Made the requested changes

karandikar committed:
Fix stylistic issues with javascript

- Rename ambiguous interval and timeout variables
- Rename functions to reduce verbosity
- Switch data structure for video dimension to be dictionary instead of list
- Remove unnecessary comment

karandikar commented:
@stvstnfrd Made a bunch of changes based on your recommendations, let me know what you think!

stvstnfrd commented:
+1 lgtm, thanks for implementing those changes!

caesar2164 commented:
@karandikar awesome, time to squash commits!

karandikar commented:
@caesar2164 @stvstnfrd Squashed, take a look and let me know if anything needs changing!

karandikar committed:
Bump release patch version from 0.1.4 to 0.1.5
